system/jlib/jmutex.hpp

```diff
@@ -661,35 +710,74 @@ class jlib_decl ReadWriteLock
     void lockWrite()        { lockWrite(false, 0); }
     bool lockRead(unsigned timeout) { return lockRead(true, timeout); }
     bool lockWrite(unsigned timeout) { return lockWrite(true, timeout); }
-    void unlock()           { 
-                                cs.enter(); 
-                                if (readLocks) readLocks--;
-                                else
-                                {
-                                    writeLocks--;
+    void changeToWrite()
+    {
+        changeToWrite(false, 0);
+    }
+    bool changeToWrite(unsigned timeout)
+    {
+        return changeToWrite(true, timeout);
+    }
+    bool changeToRead()
+    {
+        cs.enter();
+        if (!writeLocks)
+        {
+            cs.leave();
+            return false; // failed, changeToRead must be called when writeLocked
+        }
+        --writeLocks;
+        dbgassertex(!readLocks);
+        readLocks++;
+        cs.leave();
 #ifdef _DEBUG
-                                    exclWriteOwner = 0;
-#endif
-                                }
-                                assertex(writeLocks == 0);
-                                if (readLocks == 0)
-                                {
-                                    if (readWaiting)
-                                    {
-                                        unsigned numWaiting = readWaiting;
-                                        readWaiting = 0;
-                                        readLocks += numWaiting;
-                                        readSem.signal(numWaiting);
-                                    }
-                                    else if (writeWaiting)
-                                    {
-                                        writeWaiting--;
-                                        writeLocks++;
-                                        writeSem.signal();
-                                    }
-                                }
-                                cs.leave();
-                            }
+        exclWriteOwner = 0;
+#endif
+        return true;
+    }
+    void unlock()
+    {
+        cs.enter(); 
+        if (readLocks)
```

Comment:
Could do with some comments.  E.g. if (readLocks is non zero then a read lock must be being released)

Reply:
will add some comments.